<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content='A collection of cancer transcriptomics gene signatures as well as a 
    simple and tidy interface to compute single sample enrichment scores either 
    with the original procedure or with three alternatives:
    the "combined z-score" of Lee et al. (2008) &lt;doi:10.1371/journal.pcbi.1000217&gt;,
    the "single sample GSEA" of Barbie et al. (2009) &lt;doi:10.1038/nature08460&gt; and 
    the "singscore" of Foroutan et al. (2018) &lt;doi:10.1186/s12859-018-2435-4&gt;.
    The get_sig_info() function can be used to retrieve information about each 
    signature implemented.'>
<title>A Tidy Framework to Hack Gene Expression Signatures • hacksig</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Tidy Framework to Hack Gene Expression Signatures">
<meta property="og:description" content='A collection of cancer transcriptomics gene signatures as well as a 
    simple and tidy interface to compute single sample enrichment scores either 
    with the original procedure or with three alternatives:
    the "combined z-score" of Lee et al. (2008) &lt;doi:10.1371/journal.pcbi.1000217&gt;,
    the "single sample GSEA" of Barbie et al. (2009) &lt;doi:10.1038/nature08460&gt; and 
    the "singscore" of Foroutan et al. (2018) &lt;doi:10.1186/s12859-018-2435-4&gt;.
    The get_sig_info() function can be used to retrieve information about each 
    signature implemented.'>
<meta property="og:image" content="https://acare.github.io/hacksig/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">hacksig</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/hacksig.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/available-sigs.html">Available signatures</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Acare/hacksig/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="hacksig-">hacksig <a class="anchor" aria-label="anchor" href="#hacksig-"></a>
</h1>
</div>
<!-- badges: start -->

<!-- badges: end -->
<p>The goal of <code>hacksig</code> is to provide a simple and tidy interface to compute single sample scores for gene signatures and methods applied in cancer transcriptomics.</p>
<p>Scores can be obtained either for custom lists of genes or for a manually curated collection of gene signatures, including:</p>
<ul>
<li>
<a href="https://doi.org/10.1038/nm.2174" class="external-link">CINSARC</a>;</li>
<li>
<a href="https://doi.org/10.1038/ncomms3612" class="external-link">ESTIMATE</a>;</li>
<li>
<a href="https://doi.org/10.1016/j.celrep.2016.12.019" class="external-link">Immunophenoscore</a>;</li>
<li>
<a href="https://doi.org/10.1016/j.cell.2014.12.033" class="external-link">Cytolitic activity</a>;</li>
<li>and more (use <code><a href="reference/get_sig_info.html">get_sig_info()</a></code> to get a complete list of the implemented signatures)</li>
</ul>
<p>At present, signature scores can be obtained either with the original publication method or using one of three single sample scoring alternatives, namely: <em>combined z-score</em>, <em>single sample GSEA</em> and <em>singscore</em>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the last stable version of hacksig from <a href="https://CRAN.R-project.org" class="external-link">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"hacksig"</span><span class="op">)</span></span></code></pre></div>
<p>Or the development version from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Acare/hacksig"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>You can learn more about usage of the package in <code><a href="articles/hacksig.html">vignette("hacksig")</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Acare/hacksig" class="external-link">hacksig</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="available-signatures">Available signatures<a class="anchor" aria-label="anchor" href="#available-signatures"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/get_sig_info.html">get_sig_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 40 × 4</span></span>
<span><span class="co">#&gt;   signature_id       signature_keywords                          publi…¹ descr…²</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;                                       &lt;chr&gt;   &lt;chr&gt;  </span></span>
<span><span class="co">#&gt; 1 ayers2017_immexp   ayers2017_immexp|immune expanded            10.117… Immune…</span></span>
<span><span class="co">#&gt; 2 bai2019_immune     bai2019_immune|head and neck squamous cell… 10.115… Immune…</span></span>
<span><span class="co">#&gt; 3 cinsarc            cinsarc|metastasis|sarcoma|sts              10.103… Biomar…</span></span>
<span><span class="co">#&gt; 4 dececco2014_int172 dececco2014_int172|head and neck squamous … 10.109… Signat…</span></span>
<span><span class="co">#&gt; 5 eschrich2009_rsi   eschrich2009_rsi|radioresistance|radiosens… 10.101… Genes …</span></span>
<span><span class="co">#&gt; # … with 35 more rows, and abbreviated variable names ¹​publication_doi,</span></span>
<span><span class="co">#&gt; #   ²​description</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="check-your-signatures">Check your signatures<a class="anchor" aria-label="anchor" href="#check-your-signatures"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/check_sig.html">check_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, signatures <span class="op">=</span> <span class="st">"estimate"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 5</span></span>
<span><span class="co">#&gt;   signature_id     n_genes n_present frac_present missing_genes</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt; &lt;list&gt;       </span></span>
<span><span class="co">#&gt; 1 estimate_stromal     141        91        0.645 &lt;chr [50]&gt;   </span></span>
<span><span class="co">#&gt; 2 estimate_immune      141        74        0.525 &lt;chr [67]&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compute-single-sample-scores">Compute single sample scores<a class="anchor" aria-label="anchor" href="#compute-single-sample-scores"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, signatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ifng"</span>, <span class="st">"cinsarc"</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 20 × 3</span></span>
<span><span class="co">#&gt;   sample_id cinsarc muro2016_ifng</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;       &lt;dbl&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 sample1   -0.482         -0.511</span></span>
<span><span class="co">#&gt; 2 sample10  -0.0926        -1.60 </span></span>
<span><span class="co">#&gt; 3 sample11   0.730         -1.03 </span></span>
<span><span class="co">#&gt; 4 sample12  -0.625          0.851</span></span>
<span><span class="co">#&gt; 5 sample13   0.930         -0.369</span></span>
<span><span class="co">#&gt; # … with 15 more rows</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="stratify-your-samples">Stratify your samples<a class="anchor" aria-label="anchor" href="#stratify-your-samples"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_expr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="st">"estimate"</span>, method <span class="op">=</span> <span class="st">"singscore"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="reference/stratify_sig.html">stratify_sig</a></span><span class="op">(</span>cutoff <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 20 × 3</span></span>
<span><span class="co">#&gt;   sample_id estimate_immune estimate_stromal</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt;           </span></span>
<span><span class="co">#&gt; 1 sample1   low             low             </span></span>
<span><span class="co">#&gt; 2 sample10  high            high            </span></span>
<span><span class="co">#&gt; 3 sample11  high            low             </span></span>
<span><span class="co">#&gt; 4 sample12  high            low             </span></span>
<span><span class="co">#&gt; 5 sample13  low             low             </span></span>
<span><span class="co">#&gt; # … with 15 more rows</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="speed-up-computation-time">Speed-up computation time<a class="anchor" aria-label="anchor" href="#speed-up-computation-time"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, method <span class="op">=</span> <span class="st">"ssgsea"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: ℹ No genes are present in 'expr_data' for the following signatures:</span></span>
<span><span class="co">#&gt; ✖ zhu2021_ferroptosis</span></span>
<span><span class="co">#&gt; ✖ rooney2015_cyt</span></span>
<span><span class="co">#&gt; # A tibble: 20 × 39</span></span>
<span><span class="co">#&gt;   sample_id ayers2017_…¹ bai20…² cinsarc decec…³ eschr…⁴ estim…⁵ estim…⁶ eusta…⁷</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;            &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 sample1         -3914.   2316.   -13.5   1288.   1678.   -636.    778.    49.4</span></span>
<span><span class="co">#&gt; 2 sample10         1077.    575.   801.     811.   2288.   1590.   1297.  1556. </span></span>
<span><span class="co">#&gt; 3 sample11          501.   -490.  1340.    1244.   1389.   2040.    512.  -210. </span></span>
<span><span class="co">#&gt; 4 sample12         2315.   1034.  -151.     981.   3846.   1835.    772.  2138. </span></span>
<span><span class="co">#&gt; 5 sample13        -2179.    327.  1737.    1288.    665.    632.    778.  2249. </span></span>
<span><span class="co">#&gt; # … with 15 more rows, 30 more variables: fan2021_ferroptosis &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   fang2021_irgs &lt;dbl&gt;, han2021_ferroptosis &lt;dbl&gt;, he2021_ferroptosis_a &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   he2021_ferroptosis_b &lt;dbl&gt;, hu2021_derbp &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   huang2022_ferroptosis &lt;dbl&gt;, ips_cp &lt;dbl&gt;, ips_ec &lt;dbl&gt;, ips_mhc &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   ips_sc &lt;dbl&gt;, li2021_ferroptosis_a &lt;dbl&gt;, li2021_ferroptosis_b &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   li2021_ferroptosis_c &lt;dbl&gt;, li2021_ferroptosis_d &lt;dbl&gt;, li2021_irgs &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   liu2020_immune &lt;dbl&gt;, liu2021_mgs &lt;dbl&gt;, lohavanichbutr2013_hpvneg &lt;dbl&gt;, …</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>If you have any suggestions about adding new features or signatures to <code>hacksig</code>, please create an issue on <a href="https://github.com/Acare/hacksig/issues" class="external-link">GitHub</a>. Gene-level information about gene signatures are stored in <code>data-raw/hacksig_signatures.csv</code> and can be used as a template for requests.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=hacksig" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/Acare/hacksig/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Acare/hacksig/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing hacksig</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Andrea Carenzo <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-6804-9984" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Loris De Cecco <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7066-473X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Federico Pistore <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3254-9247" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://CRAN.R-project.org/package=hacksig" class="external-link"><img src="https://www.r-pkg.org/badges/version/hacksig" alt="CRAN status"></a></li>
<li><a href="https://app.codecov.io/gh/Acare/hacksig?branch=master" class="external-link"><img src="https://codecov.io/gh/Acare/hacksig/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://github.com/Acare/hacksig/actions" class="external-link"><img src="https://github.com/Acare/hacksig/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Andrea Carenzo, Loris De Cecco, Federico Pistore.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
