---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
options(tibble.print_min = 5L, tibble.print_max = 5L)
```

# hacksig

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![CRAN status](https://www.r-pkg.org/badges/version/hacksig)](https://CRAN.R-project.org/package=hacksig)
[![Codecov test coverage](https://codecov.io/gh/Acare/hacksig/branch/master/graph/badge.svg)](https://app.codecov.io/gh/Acare/hacksig?branch=master)
<!-- badges: end -->

The goal of `hacksig` is to provide a simple and tidy interface to compute single 
sample scores for a manually curated collection of gene signatures and methods 
applied in cancer transcriptomics.

Several gene signatures and methods are implemented, including:

  * [CINSARC](https://doi.org/10.1038/nm.2174);
  * [ESTIMATE](https://doi.org/10.1038/ncomms3612);
  * [Immunophenoscore](https://doi.org/10.1016/j.celrep.2016.12.019);
  * [Cytolitic activity](https://doi.org/10.1016/j.cell.2014.12.033);
  * and many more (use `get_sig_info()` for a complete list of gene signatures implemented)

One can choose to apply either the original publication method or one of three 
single sample scoring alternatives, namely: combined z-score, single sample GSEA 
and singscore.

## Installation

```{r, echo=FALSE, eval=FALSE}
# You can install the released version of hacksig from [CRAN](https://CRAN.R-project.org) with:
# 
# install.packages("hacksig")
```

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("Acare/hacksig")
```

## Usage

To show example usage of the package, we will use a simulated gene expression matrix `test_expr`, which is stored as an R object in `hacksig`.

```{r, message=FALSE}
library(hacksig)
library(dplyr)
library(future)
```

### Get available signatures

```{r}
get_sig_info()
```

### Check your signatures

```{r}
check_sig(test_expr, signatures = "immune")
```

### Compute single sample scores

```{r}
hack_sig(test_expr, signatures = c("estimate", "cinsarc"), method = "zscore")
```

### Classify your samples

```{r}
test_expr %>% 
    hack_sig("estimate", method = "singscore", direction = "up") %>% 
    hack_class()
```

### Speed-up computation time

```{r}
plan(multisession)
hack_sig(test_expr, method = "ssgsea")
```


## Contributing

If you have any suggestions about adding new features to `hacksig`, please 
open an issue request on [GitHub](https://github.com/Acare/hacksig/issues).
Gene-level information about gene signatures are stored in the R object 
`signatures_data` and can be used as a template for requests.

