---
title: "Introduction to hacksig"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to hacksig}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(hacksig)
```

This document will show you how to getting started with hacksig.
We will use `test_expr` as a gene expression matrix
We must load also these packages.

```{r, message=FALSE}
library(dplyr)
library(future)
library(ggplot2)
library(msigdbr)
library(purrr)
library(tibble)
library(tidyr)
```

## Getting information about the hacksig signatures

hacksig is a collection of gene signatures and methods found in the cancer 
transcriptomics literature. You can find 

```{r}
get_sig_info()
```



```{r}
# usa purrr::map() per far vedere come ottenere output per diversi valori dei parametri
# usa ggplot2 per mostrare i risultati
```

### Check your signatures

The first thing you should do before computing signature scores is to check how 
much of the genes composing a signature are present in your gene expression matrix.
To accomplish this, we can use `check_sig()`:

```{r}
check_sig(test_expr)
```

By default, `check_sig()` will compute statistics for every signature implemented 
in `hacksig`. You can filter for specific signatures by entering a keyword in the 
`signatures` argument:

```{r}
check_sig(test_expr, signatures = "immune")
```

The set of available keywords associated to gene signatures can be retrieved with:

```{r}
get_sig_info()
```

We can also check for signatures not implemented in hacksig, that is custom signatures.
For example, we can use the `msigdbr` package to download the *Hallmark* gene set collection.

```{r}
hallmark_list <- msigdbr(species = "Homo sapiens", category = "H") %>%
    distinct(gs_name, gene_symbol) %>%
    nest(genes = c(gene_symbol)) %>%
    mutate(genes = map(genes, compose(as_vector, unname))) %>%
    deframe()
check_sig(test_expr, hallmark_list)
```

### Compute single sample scores

The main function of the package is `hack_sig()` and it permits to obtain single 
sample scores from gene signatures.
By default, `hack_sig()` will compute scores for all the signatures implemented 
in the package with the original publication method (e.g. weighted sum of expression 
values and model coefficients).

```{r}
hack_sig(test_expr)
```

You can also filter for specific signatures (e.g. the immune and stromal ESTIMATE 
signatures) and choose a particular single sample method:

```{r}
hack_sig(test_expr, signatures = "estimate", method = "zscore")
```

Valid choices for single sample `method`s are:

  * `zscore`, for the combined z-score;
  * `ssgsea`, for the single sample GSEA;
  * `singscore`, for the singscore method.
  
Run `?hack_sig` to see references for these methods.

As in `check_sig()`, the argument `signatures` can also be a list of gene signatures.
For example, we can compute normalized single sample GSEA scores for the Hallmark gene sets:

```{r}
hack_sig(test_expr, hallmark_list, 
         method = "ssgsea", sample_norm = "separate", alpha = 0.5)
```

There are three methods for which `hack_sig()` cannot be used to compute gene signature
scores with the original method. These are: CINSARC, ESTIMATE and the Immunophenoscore.

For the CINSARC classification, you must provide a vector with distant metastasis status:

```{r}
rand_dm <- sample(c(0, 1), size = ncol(test_expr), replace = TRUE)
hack_cinsarc(test_expr, rand_dm)
```

Immune, stromal, ESTIMATE and tumor purity scores from the ESTIMATE method can be obtained with: 

```{r}
hack_estimate(test_expr)
```

Finally, the raw immunophenoscore and its discrete counterpart can be obtained with:

```{r}
hack_immunophenoscore(test_expr)
```


### Classify your samples

If you want to categorize your samples into two or more signature classes based 
on a score cutoff, you can use `hack_class()` after `hack_sig()`:

```{r}
test_expr %>% 
    hack_sig("estimate", method = "singscore", direction = "up") %>% 
    hack_class()
```

### Speed-up computation time

Our single sample method implementations based on ranks (i.e. single sample GSEA and singscore) are slower than their counterparts implemented in `GSVA` and `singscore`.
Hence, to speed-up computation time you can use the `future` package:

```{r}
plan(multisession)
hack_sig(test_expr, hallmark_list, method = "ssgsea")
```
