<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Hack the CINSARC classification — hack_cinsarc • hacksig</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Hack the CINSARC classification — hack_cinsarc"><meta name="description" content="Given a gene expression matrix and a 0-1 vector indicating the distant metastasis
status of samples, hack_cinsarc() classifies samples into one of two risk
classes, C1 or C2, using the CINSARC signature as implemented in
Chibon et al., 2010."><meta property="og:description" content="Given a gene expression matrix and a 0-1 vector indicating the distant metastasis
status of samples, hack_cinsarc() classifies samples into one of two risk
classes, C1 or C2, using the CINSARC signature as implemented in
Chibon et al., 2010."><meta property="og:image" content="https://acare.github.io/hacksig/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hacksig</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/hacksig.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/available-sigs.html">Available signatures</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Acare/hacksig/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Hack the CINSARC classification</h1>
      <small class="dont-index">Source: <a href="https://github.com/Acare/hacksig/blob/master/R/hack_cinsarc.R" class="external-link"><code>R/hack_cinsarc.R</code></a></small>
      <div class="d-none name"><code>hack_cinsarc.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given a gene expression matrix and a 0-1 vector indicating the distant metastasis
status of samples, <code>hack_cinsarc()</code> classifies samples into one of two risk
classes, C1 or C2, using the CINSARC signature as implemented in
<em>Chibon et al., 2010</em>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">hack_cinsarc</span><span class="op">(</span><span class="va">expr_data</span>, <span class="va">dm_status</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p><a href="https://codeocean.com/capsule/4933686/tree/v4" class="external-link">codeocean.com/capsule/4933686/tree/v4</a></p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-expr-data">expr_data<a class="anchor" aria-label="anchor" href="#arg-expr-data"></a></dt>
<dd><p>A normalized gene expression matrix (or data frame) with
gene symbols as row names and samples as columns.</p></dd>


<dt id="arg-dm-status">dm_status<a class="anchor" aria-label="anchor" href="#arg-dm-status"></a></dt>
<dd><p>A numeric vector specifying whether a sample has either (1)
or not (0) developed distant metastasis.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A tibble with one row for each sample in <code>expr_data</code> and two columns:
<code>sample_id</code> and <code>cinsarc_class</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>CINSARC (<em>Complexity INdex in SARComas</em>) is a prognostic 67-gene signature
related to mitosis and control of chromosome integrity.
It was developed to improve metastatic outcome prediction in soft tissue
sarcomas over the FNCLCC (<em>Fédération Francaise des Centres de Lutte Contre
le Cancer</em>) grading system.</p>
    </div>
    <div class="section level2">
    <h2 id="algorithm">Algorithm<a class="anchor" aria-label="anchor" href="#algorithm"></a></h2>


<p>The CINSARC method implemented in <code>hacksig</code> makes use of leave-one-out cross
validation (LOOCV) to classify samples into C1/C2 risk groups (see <em>Lesluyes &amp; Chibon, 2020</em>).
First, gene expression values are centered by their mean across samples.
Then, for each iteration of the LOOCV, mean normalized gene values are computed
by metastasis group (i.e. compute the metastatic centroids). Then, one minus the
Spearman's correlation between centered samples and metastatic centroids are computed.
Finally, if a sample is more correlated to the non-metastatic centroid, then
it is assigned to the C1 class (low risk). Conversely, if a sample is more
correlated to the metastatic centroid, then it is assigned to the C2 class (high risk).</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Chibon, F., Lagarde, P., Salas, S., Pérot, G., Brouste, V., Tirode, F.,
Lucchesi, C., de Reynies, A., Kauffmann, A., Bui, B., Terrier, P.,
Bonvalot, S., Le Cesne, A., Vince-Ranchère, D., Blay, J. Y., Collin, F.,
Guillou, L., Leroux, A., Coindre, J. M., &amp; Aurias, A. (2010). Validated
prediction of clinical outcome in sarcomas and multiple types of cancer on
the basis of a gene expression signature related to genome complexity.
<em>Nature medicine</em>, 16(7), 781–787. <a href="https://doi.org/10.1038/nm.2174" class="external-link">doi:10.1038/nm.2174</a>
.</p>
<p>Lesluyes, T., &amp; Chibon, F. (2020). A Global and Integrated
Analysis of CINSARC-Associated Genetic Defects. <em>Cancer research</em>, 80(23),
5282–5290. <a href="https://doi.org/10.1158/0008-5472.CAN-20-0512" class="external-link">doi:10.1158/0008-5472.CAN-20-0512</a>
.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># generate random distant metastasis outcome</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">test_dm_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">hack_cinsarc</span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">test_dm_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 20 × 2</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sample_id cinsarc_class</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> sample1   C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> sample2   C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> sample3   C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> sample4   C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> sample5   C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> sample6   C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> sample7   C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> sample8   C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> sample9   C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> sample10  C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> sample11  C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> sample12  C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> sample13  C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> sample14  C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> sample15  C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> sample16  C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> sample17  C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> sample18  C2           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">19</span> sample19  C1           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">20</span> sample20  C2           </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrea Carenzo, Loris De Cecco, Federico Pistore.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

