<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Score samples by gene signatures — hack_sig • hacksig</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Score samples by gene signatures — hack_sig"><meta name="description" content="Compute gene signature single sample scores in one of different ways.
You can choose to apply either the original procedure or one of three single
sample scoring methods: the combined z-score (Lee et al., 2008), the single
sample GSEA (Barbie et al., 2009) or the singscore method (Foroutan et al., 2018)."><meta property="og:description" content="Compute gene signature single sample scores in one of different ways.
You can choose to apply either the original procedure or one of three single
sample scoring methods: the combined z-score (Lee et al., 2008), the single
sample GSEA (Barbie et al., 2009) or the singscore method (Foroutan et al., 2018)."><meta property="og:image" content="https://acare.github.io/hacksig/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hacksig</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/hacksig.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/available-sigs.html">Available signatures</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Acare/hacksig/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Score samples by gene signatures</h1>
      <small class="dont-index">Source: <a href="https://github.com/Acare/hacksig/blob/master/R/hack_sig.R" class="external-link"><code>R/hack_sig.R</code></a></small>
      <div class="d-none name"><code>hack_sig.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compute gene signature single sample scores in one of different ways.
You can choose to apply either the <em>original</em> procedure or one of three single
sample scoring methods: the combined z-score (<em>Lee et al., 2008</em>), the single
sample GSEA (<em>Barbie et al., 2009</em>) or the singscore method (<em>Foroutan et al., 2018</em>).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">hack_sig</span><span class="op">(</span></span>
<span>  <span class="va">expr_data</span>,</span>
<span>  signatures <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"original"</span>,</span>
<span>  direction <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  sample_norm <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>  rank_norm <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.25</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-expr-data">expr_data<a class="anchor" aria-label="anchor" href="#arg-expr-data"></a></dt>
<dd><p>A normalized gene expression matrix (or data frame) with
gene symbols as row names and samples as columns.</p></dd>


<dt id="arg-signatures">signatures<a class="anchor" aria-label="anchor" href="#arg-signatures"></a></dt>
<dd><p>It can be a list of signatures or a character vector indicating
keywords for a group of signatures. The default (<code>"all"</code>) will cause the
function to compute single sample scores for all the signatures implemented
in <code>hacksig</code>.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string specifying which method to use for computing
the single sample score for each signature. You can choose one of:</p><ul><li><p><code>"original"</code>, the original method used by the authors of the signature;</p></li>
<li><p><code>"zscore"</code>, the combined z-score method;</p></li>
<li><p><code>"ssgsea"</code>, the single sample GSEA method;</p></li>
<li><p><code>"singscore"</code>, the singscore method;</p></li>
</ul></dd>


<dt id="arg-direction">direction<a class="anchor" aria-label="anchor" href="#arg-direction"></a></dt>
<dd><p>A character string specifying the <strong>singscore</strong> computation
method depending on the direction of the signatures. Can be on of:</p><ul><li><p><code>"none"</code> (default), undirected signatures, that is you do not know whether
the genes are up- or down-regulated;</p></li>
<li><p><code>"up"</code>, all genes in the signature are supposed to be up-regulated;</p></li>
<li><p><code>"down"</code>, all genes in the signature are supposed to be down-regulated;</p></li>
</ul></dd>


<dt id="arg-sample-norm">sample_norm<a class="anchor" aria-label="anchor" href="#arg-sample-norm"></a></dt>
<dd><p>A character string specifying the type of normalization
affecting the <strong>single sample GSEA</strong> scores. Can be one of:</p><ul><li><p><code>"raw"</code> (default), obtain raw scores;</p></li>
<li><p><code>"separate"</code>, normalize raw scores in \([0, 1]\) across samples for
each signature separately.</p></li>
<li><p><code>"all"</code>, normalize raw scores both across samples and signatures.</p></li>
</ul></dd>


<dt id="arg-rank-norm">rank_norm<a class="anchor" aria-label="anchor" href="#arg-rank-norm"></a></dt>
<dd><p>A character string specifying how gene expression ranks should
be normalized in the <strong>single sample GSEA</strong> procedure. Valid choices are:</p><ul><li><p><code>"none"</code> (default), no rank normalization;</p></li>
<li><p><code>"rank"</code>, ranks are multiplied by <code>10000 / nrow(expr_data)</code>;</p></li>
<li><p><code>"logrank"</code>, normalized ranks are logged.</p></li>
</ul></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>A numeric scalar. Exponent in the running sum of the <strong>single sample GSEA</strong>
score calculation which weighs the gene ranks. Defaults to \(\alpha = 0.25\).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A tibble with one row for each sample in <code>expr_data</code>, a column <code>sample_id</code>
indicating sample identifiers and one column for each input signature giving
single sample scores.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For <code>"original"</code> method, it is intended the procedure used in the original
publication by the authors for computing the signature score.
<code>hack_sig()</code> can compute signature scores with the original method only if
this is a relatively simple procedure (e.g weighted sum of fitted model
coefficients and expression values).
For more complex methods, such as CINSARC, ESTIMATE and Immunophenoscore,
use the dedicated functions.</p>
<p>If <code>signatures</code> is a custom list of gene signatures, then the <code>"ssgsea"</code>
method will be applied by default.</p>
    </div>
    <div class="section level2">
    <h2 id="algorithm">Algorithm<a class="anchor" aria-label="anchor" href="#algorithm"></a></h2>


<p>This section gives a brief explanation of how single sample scores are obtained
from different methods.</p><div class="section">
<h3 id="combined-z-score">Combined z-score<a class="anchor" aria-label="anchor" href="#combined-z-score"></a></h3>


<p>Gene expression values are centered by their mean value and scaled by their
standard deviation across samples for each gene (z-scores). Then, for each
sample and signature, corresponding z-scores are added up and divided by the
square root of the signature size (i.e. the number of genes composing a signature).</p>
<p>The combined z-score method is also implemented in the R package <code>GSVA</code>
(<em>Hänzelmann et al., 2013</em>).</p>
</div>

<div class="section">
<h3 id="single-sample-gsea">Single sample GSEA<a class="anchor" aria-label="anchor" href="#single-sample-gsea"></a></h3>


<p>For each sample, genes are ranked by expression value in increasing order and
rank normalization may follow (see argument <code>rank_norm</code>). Then, two probability-like
vectors are computed for each sample and signature:</p><ul><li><p>\(P_{in}\), the cumulative sum of weighted ranks divided by their total
sum for genes in the signature;</p></li>
<li><p>\(P_{out}\), the cumulative sum of ones (indicating genes not in the signature)
divided by the number of genes not in the signature.</p></li>
</ul><p>The single sample GSEA score is obtained by adding up the elements of the
vector difference \(P_{in} - P_{out}\).
Finally, single sample scores could be normalized either across samples or across
gene signatures and samples.</p>
<p>The single sample GSEA method is also implemented in the R package <code>GSVA</code>
(<em>Hänzelmann et al., 2013</em>).</p>
</div>

<div class="section">
<h3 id="singscore">Singscore<a class="anchor" aria-label="anchor" href="#singscore"></a></h3>


<p>For signatures whose genes are supposed to be up- or down-regulated, genes
are ranked by expression value in increasing or decreasing order, respectively.
For signatures whose direction is unknown, genes are ranked by absolute expression
in increasing order and are median-centered.
Enrichment scores are then computed for each sample and signature by averaging
gene ranks for genes in the signature.
Finally, normalized scores are obtained by subtracting the theoretical minimum
mean rank from the score and dividing by the difference between the theoretical
maximum and minimum mean ranks.</p>
<p>The <code>hacksig</code> implementation of this method works only with unidirectional (i.e.
all genes up- or down-regulated) and undirected gene signatures.
If you want to get single sample scores for bidirectional gene signatures (i.e.
signatures composed of both up- and down-regulated genes), please use the R
package <code>singscore</code> (<em>Foroutan et al., 2018</em>).</p>
</div>

    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Barbie, D. A., Tamayo, P., Boehm, J. S., Kim, S. Y., Moody, S. E., Dunn, I. F.,
Schinzel, A. C., Sandy, P., Meylan, E., Scholl, C., Fröhling, S., Chan, E. M.,
Sos, M. L., Michel, K., Mermel, C., Silver, S. J., Weir, B. A., Reiling, J. H.,
Sheng, Q., Gupta, P. B., … Hahn, W. C. (2009). Systematic RNA interference
reveals that oncogenic KRAS-driven cancers require TBK1. <em>Nature</em>, 462(7269),
108–112. <a href="https://doi.org/10.1038/nature08460" class="external-link">doi:10.1038/nature08460</a>
.</p>
<p>Foroutan, M., Bhuva, D. D., Lyu, R., Horan, K., Cursons, J., &amp;
Davis, M. J. (2018). Single sample scoring of molecular phenotypes.
<em>BMC bioinformatics</em>, 19(1), 404. <a href="https://doi.org/10.1186/s12859-018-2435-4" class="external-link">doi:10.1186/s12859-018-2435-4</a>
.</p>
<p>Hänzelmann, S., Castelo, R., &amp; Guinney, J. (2013). GSVA: gene set variation
analysis for microarray and RNA-seq data. <em>BMC bioinformatics</em>, 14, 7.
<a href="https://doi.org/10.1186/1471-2105-14-7" class="external-link">doi:10.1186/1471-2105-14-7</a>
.</p>
<p>Lee, E., Chuang, H. Y., Kim, J. W., Ideker, T., &amp; Lee, D. (2008). Inferring
pathway activity toward precise disease classification.
<em>PLoS computational biology</em>, 4(11), e1000217.
<a href="https://doi.org/10.1371/journal.pcbi.1000217" class="external-link">doi:10.1371/journal.pcbi.1000217</a>
.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="get_sig_info.html">get_sig_info()</a></code> to get information about all implemented signatures.</p>
<p><code><a href="check_sig.html">check_sig()</a></code> to check if signatures are applicable to your data.</p>
<p><code><a href="hack_cinsarc.html">hack_cinsarc()</a></code> to apply the original CINSARC procedure.</p>
<p><code><a href="hack_estimate.html">hack_estimate()</a></code> to obtain the original ESTIMATE scores.</p>
<p><code><a href="hack_immunophenoscore.html">hack_immunophenoscore()</a></code> to apply the original Immunophenoscore procedure.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Raw ssGSEA scores for all implemented signatures can be obtained with:</span></span></span>
<span class="r-in"><span><span class="fu">hack_sig</span><span class="op">(</span><span class="va">test_expr</span>, method <span class="op">=</span> <span class="st">"ssgsea"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span><span style="color: #0000BB;">ℹ</span> No genes are present in 'expr_data' for the following signatures:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> zhu2021_ferroptosis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> rooney2015_cyt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 20 × 39</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sample_id ayers2017_immexp bai2019_immune cinsarc dececco2014_int172</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> sample1            -<span style="color: #BB0000; text-decoration: underline;">3</span><span style="color: #BB0000;">914.</span>           <span style="text-decoration: underline;">2</span>316.   -<span style="color: #BB0000;">13.5</span>              <span style="text-decoration: underline;">1</span>288.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> sample10            <span style="text-decoration: underline;">1</span>077.            575.   801.                811.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> sample11             501.           -<span style="color: #BB0000;">490.</span>  <span style="text-decoration: underline;">1</span>340.               <span style="text-decoration: underline;">1</span>244.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> sample12            <span style="text-decoration: underline;">2</span>315.           <span style="text-decoration: underline;">1</span>034.  -<span style="color: #BB0000;">151.</span>                981.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> sample13           -<span style="color: #BB0000; text-decoration: underline;">2</span><span style="color: #BB0000;">179.</span>            327.  <span style="text-decoration: underline;">1</span>737.               <span style="text-decoration: underline;">1</span>288.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> sample14            <span style="text-decoration: underline;">1</span>274.           <span style="text-decoration: underline;">2</span>551.  <span style="text-decoration: underline;">1</span>792.               <span style="text-decoration: underline;">1</span>118.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> sample15             668.           <span style="text-decoration: underline;">4</span>211.   734.                961.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> sample16           -<span style="color: #BB0000; text-decoration: underline;">1</span><span style="color: #BB0000;">014.</span>           <span style="text-decoration: underline;">1</span>283.  <span style="text-decoration: underline;">1</span>123.                867.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> sample17            -<span style="color: #BB0000;">476.</span>           <span style="text-decoration: underline;">2</span>509.   384.               <span style="text-decoration: underline;">2</span>175.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> sample18            <span style="text-decoration: underline;">1</span>101.           -<span style="color: #BB0000;">315.</span>  <span style="text-decoration: underline;">1</span>911.                985.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> sample19            <span style="text-decoration: underline;">3</span>938.           <span style="text-decoration: underline;">3</span>566.  -<span style="color: #BB0000;">220.</span>               <span style="text-decoration: underline;">1</span>959.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> sample2            -<span style="color: #BB0000; text-decoration: underline;">3</span><span style="color: #BB0000;">348.</span>           <span style="text-decoration: underline;">1</span>350. -<span style="color: #BB0000; text-decoration: underline;">1</span><span style="color: #BB0000;">070.</span>               <span style="text-decoration: underline;">1</span>322.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> sample20            <span style="text-decoration: underline;">1</span>261.            975.  <span style="text-decoration: underline;">1</span>442.                703.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> sample3             <span style="text-decoration: underline;">1</span>697.           <span style="text-decoration: underline;">1</span>829.  <span style="text-decoration: underline;">1</span>805.                685.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> sample4              366.           <span style="text-decoration: underline;">5</span>611.   326.               <span style="text-decoration: underline;">1</span>684.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> sample5              969.           <span style="text-decoration: underline;">1</span>224.   290.                718.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> sample6             <span style="text-decoration: underline;">3</span>820.           -<span style="color: #BB0000;">507.</span>   782.                382.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> sample7              920.           <span style="text-decoration: underline;">3</span>452.  <span style="text-decoration: underline;">1</span>835.               <span style="text-decoration: underline;">1</span>230.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">19</span> sample8              -<span style="color: #BB0000;">13.8</span>          -<span style="color: #BB0000;">229.</span>   462.                966.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">20</span> sample9             <span style="text-decoration: underline;">1</span>819.           <span style="text-decoration: underline;">3</span>073.   824.                525.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 34 more variables: eschrich2009_rsi &lt;dbl&gt;, estimate_immune &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   estimate_stromal &lt;dbl&gt;, eustace2013_hypoxia &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   fan2021_ferroptosis &lt;dbl&gt;, fang2021_irgs &lt;dbl&gt;, han2021_ferroptosis &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   he2021_ferroptosis_a &lt;dbl&gt;, he2021_ferroptosis_b &lt;dbl&gt;, hu2021_derbp &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   huang2022_ferroptosis &lt;dbl&gt;, ips_cp &lt;dbl&gt;, ips_ec &lt;dbl&gt;, ips_mhc &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ips_sc &lt;dbl&gt;, li2021_ferroptosis_a &lt;dbl&gt;, li2021_ferroptosis_b &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   li2021_ferroptosis_c &lt;dbl&gt;, li2021_ferroptosis_d &lt;dbl&gt;, …</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To obtain 0-1 normalized ssGSEA scores, use:</span></span></span>
<span class="r-in"><span><span class="fu">hack_sig</span><span class="op">(</span><span class="va">test_expr</span>, method <span class="op">=</span> <span class="st">"ssgsea"</span>, sample_norm <span class="op">=</span> <span class="st">"separate"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span><span style="color: #0000BB;">ℹ</span> No genes are present in 'expr_data' for the following signatures:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> zhu2021_ferroptosis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> rooney2015_cyt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 20 × 39</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sample_id ayers2017_immexp bai2019_immune cinsarc dececco2014_int172</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> sample1             0             0.461     0.354             0.505 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> sample10            0.636         0.177     0.628             0.239 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> sample11            0.562         0.002<span style="text-decoration: underline;">73</span>   0.808             0.481 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> sample12            0.793         0.252     0.308             0.334 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> sample13            0.221         0.136     0.942             0.505 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> sample14            0.661         0.500     0.960             0.411 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> sample15            0.584         0.771     0.605             0.323 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> sample16            0.369         0.293     0.736             0.271 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> sample17            0.438         0.493     0.488             1     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> sample18            0.639         0.031<span style="text-decoration: underline;">4</span>    1                 0.337 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> sample19            1             0.666     0.285             0.880 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> sample2             0.072<span style="text-decoration: underline;">1</span>        0.303     0                 0.524 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> sample20            0.659         0.242     0.842             0.179 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> sample3             0.715         0.382     0.964             0.169 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> sample4             0.545         1         0.468             0.726 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> sample5             0.622         0.283     0.456             0.188 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> sample6             0.985         0         0.621             0     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> sample7             0.616         0.647     0.974             0.473 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">19</span> sample8             0.497         0.045<span style="text-decoration: underline;">4</span>    0.514             0.326 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">20</span> sample9             0.730         0.585     0.635             0.079<span style="text-decoration: underline;">6</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 34 more variables: eschrich2009_rsi &lt;dbl&gt;, estimate_immune &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   estimate_stromal &lt;dbl&gt;, eustace2013_hypoxia &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   fan2021_ferroptosis &lt;dbl&gt;, fang2021_irgs &lt;dbl&gt;, han2021_ferroptosis &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   he2021_ferroptosis_a &lt;dbl&gt;, he2021_ferroptosis_b &lt;dbl&gt;, hu2021_derbp &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   huang2022_ferroptosis &lt;dbl&gt;, ips_cp &lt;dbl&gt;, ips_ec &lt;dbl&gt;, ips_mhc &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ips_sc &lt;dbl&gt;, li2021_ferroptosis_a &lt;dbl&gt;, li2021_ferroptosis_b &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   li2021_ferroptosis_c &lt;dbl&gt;, li2021_ferroptosis_d &lt;dbl&gt;, …</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># You can also change the exponent of the ssGSEA running sum with:</span></span></span>
<span class="r-in"><span><span class="fu">hack_sig</span><span class="op">(</span><span class="va">test_expr</span>, method <span class="op">=</span> <span class="st">"ssgsea"</span>, sample_norm <span class="op">=</span> <span class="st">"separate"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span><span style="color: #0000BB;">ℹ</span> No genes are present in 'expr_data' for the following signatures:</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> zhu2021_ferroptosis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> rooney2015_cyt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 20 × 39</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sample_id ayers2017_immexp bai2019_immune cinsarc dececco2014_int172</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> sample1             0              0.684    0.263            0.487  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> sample10            0.741          0.186    0.659            0.316  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> sample11            0.574          0.048<span style="text-decoration: underline;">6</span>   0.811            0.562  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> sample12            0.779          0.280    0.310            0.394  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> sample13            0.268          0.112    1                0.502  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> sample14            0.705          0.503    0.960            0.568  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> sample15            0.674          0.746    0.625            0.507  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> sample16            0.461          0.299    0.770            0.298  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> sample17            0.443          0.481    0.424            1      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> sample18            0.699          0.022<span style="text-decoration: underline;">3</span>   0.987            0.403  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> sample19            1              0.716    0.318            0.890  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> sample2             0.073<span style="text-decoration: underline;">2</span>         0.325    0                0.537  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> sample20            0.730          0.399    0.886            0.128  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> sample3             0.747          0.363    0.980            0.153  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> sample4             0.555          1        0.468            0.838  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> sample5             0.639          0.360    0.501            0.220  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> sample6             0.966          0        0.693            0      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> sample7             0.622          0.644    0.985            0.466  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">19</span> sample8             0.498          0.090<span style="text-decoration: underline;">2</span>   0.525            0.395  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">20</span> sample9             0.825          0.584    0.557            0.009<span style="text-decoration: underline;">33</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 34 more variables: eschrich2009_rsi &lt;dbl&gt;, estimate_immune &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   estimate_stromal &lt;dbl&gt;, eustace2013_hypoxia &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   fan2021_ferroptosis &lt;dbl&gt;, fang2021_irgs &lt;dbl&gt;, han2021_ferroptosis &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   he2021_ferroptosis_a &lt;dbl&gt;, he2021_ferroptosis_b &lt;dbl&gt;, hu2021_derbp &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   huang2022_ferroptosis &lt;dbl&gt;, ips_cp &lt;dbl&gt;, ips_ec &lt;dbl&gt;, ips_mhc &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   ips_sc &lt;dbl&gt;, li2021_ferroptosis_a &lt;dbl&gt;, li2021_ferroptosis_b &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   li2021_ferroptosis_c &lt;dbl&gt;, li2021_ferroptosis_d &lt;dbl&gt;, …</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To obtain combined z-scores for custom gene signatures, use:</span></span></span>
<span class="r-in"><span><span class="va">custom_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rand_sig1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                    rand_sig2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span>, <span class="st">"RANDOMGENE"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">hack_sig</span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">custom_list</span>, method <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 20 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sample_id rand_sig1 rand_sig2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> sample1      0.601     0.160 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> sample10    -<span style="color: #BB0000;">0.907</span>    -<span style="color: #BB0000;">0.440</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> sample11    -<span style="color: #BB0000;">1.06</span>     -<span style="color: #BB0000;">0.071</span><span style="color: #BB0000; text-decoration: underline;">7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> sample12    -<span style="color: #BB0000;">1.46</span>     -<span style="color: #BB0000;">0.050</span><span style="color: #BB0000; text-decoration: underline;">4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> sample13    -<span style="color: #BB0000;">0.177</span>    -<span style="color: #BB0000;">0.204</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> sample14    -<span style="color: #BB0000;">0.484</span>    -<span style="color: #BB0000;">2.43</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> sample15     0.868     0.057<span style="text-decoration: underline;">7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> sample16    -<span style="color: #BB0000;">0.266</span>     1.94  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> sample17    -<span style="color: #BB0000;">0.146</span>    -<span style="color: #BB0000;">0.041</span><span style="color: #BB0000; text-decoration: underline;">0</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> sample18     0.456     0.895 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> sample19    -<span style="color: #BB0000;">0.052</span><span style="color: #BB0000; text-decoration: underline;">2</span>   -<span style="color: #BB0000;">0.142</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> sample2     -<span style="color: #BB0000;">2.03</span>     -<span style="color: #BB0000;">0.230</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> sample20     0.258     0.879 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> sample3      1.27      1.65  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> sample4      0.899    -<span style="color: #BB0000;">0.498</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> sample5      0.963    -<span style="color: #BB0000;">0.520</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">17</span> sample6      1.61      0.703 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">18</span> sample7     -<span style="color: #BB0000;">1.11</span>     -<span style="color: #BB0000;">1.11</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">19</span> sample8      2.11     -<span style="color: #BB0000;">0.190</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">20</span> sample9     -<span style="color: #BB0000;">1.35</span>     -<span style="color: #BB0000;">0.364</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrea Carenzo, Loris De Cecco, Federico Pistore.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

