<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Available signatures â€¢ hacksig</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Available signatures">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">hacksig</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/hacksig.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/available-sigs.html">Available signatures</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Acare/hacksig/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="available-sigs_files/core-js-2.5.3/shim.min.js"></script><script src="available-sigs_files/react-18.2.0/react.min.js"></script><script src="available-sigs_files/react-18.2.0/react-dom.min.js"></script><script src="available-sigs_files/reactwidget-2.0.0/react-tools.js"></script><link href="available-sigs_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="available-sigs_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="available-sigs_files/reactable-0.4.4/reactable.css" rel="stylesheet">
<script src="available-sigs_files/reactable-binding-0.4.4/reactable.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Available signatures</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Acare/hacksig/blob/master/vignettes/available-sigs.Rmd" class="external-link"><code>vignettes/available-sigs.Rmd</code></a></small>
      <div class="d-none name"><code>available-sigs.Rmd</code></div>
    </div>

    
    
<p>The following table is based on the output of
<code><a href="../reference/get_sig_info.html">get_sig_info()</a></code> and it can be used to explore in a simple
way all the implemented gene signatures in hacksig.</p>
<p>The search bar allows you to search for specific type of signatures
and links in the DOI column redirect to the original papers.</p>
<div class="reactable html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"signature_id":["ayers2017_immexp","bai2019_immune","cinsarc","dececco2014_int172","eschrich2009_rsi","estimate_immune","estimate_stromal","eustace2013_hypoxia","fan2021_ferroptosis","fang2021_irgs","han2021_ferroptosis","he2021_ferroptosis_a","he2021_ferroptosis_b","hu2021_derbp","huang2022_ferroptosis","ips_cp","ips_ec","ips_mhc","ips_sc","li2021_ferroptosis_a","li2021_ferroptosis_b","li2021_ferroptosis_c","li2021_ferroptosis_d","li2021_irgs","liu2020_immune","liu2021_mgs","lohavanichbutr2013_hpvneg","lu2020_npc","lu2021_ferroptosis","muro2016_ifng","qiang2021_irgs","rooney2015_cyt","shan2021_ferroptosis","she2020_irgs","si2022_npc","tang2018_npc","wu2020_metabolic","xu2021_ferroptosis","zhu2021_ferroptosis","zou2020_npc"],"signature_keywords":["ayers2017_immexp|immune expanded","bai2019_immune|head and neck squamous cell carcinoma|hnscc|inflammation|ocscc|oral cavity","cinsarc|metastasis|sarcoma|sts","dececco2014_int172|head and neck squamous cell carcinoma|hnscc","eschrich2009_rsi|radioresistance|radiosensitivity|rt","estimate|immune","estimate|stromal","eustace2013_hypoxia|head and neck squamous cell carcinoma|hnscc|larynx","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|mutation|prognosis","fang2021_irgs|immune","ferroptosis|gene expression omnibus|geo|head and neck squamous cell carcinoma|hnscc|laryngeal cancer|larynx|prognosis|tcga|the cancer genome atlas","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis","head and neck squamous cell carcinoma|hnscc|hu2021_derbp|rna-binding protein","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis|therapeutic targets","cd27|checkpoint|ctla4|icos|ido1|immune|immunophenoscore|ips_cp|lag3|pd1|pdl1|pdl2|tigit|tim3","act cd4|act cd8|immune|immunophenoscore|ips_ec|tem cd4|tem cd8","b2m|hla-a|hla-b|hla-c|hla-dpa1|hla-dpb1|hla-e|hla-f|immune|immunophenoscore|ips_mhc|tap1|tap2","immune|immunophenoscore|ips_sc|mdsc|treg","ferroptosis|head and neck squamous cell carcinoma|hnscc|ocscc|oral cavity|prognosis","ferroptosis|head and neck squamous cell carcinoma|hnscc|prognosis|tcga|the cancer genome atlas","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis|tcga|the cancer genome atlas","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis|tcga|the cancer genome atlas","immune exhaustion|li2021_irgs|thyroid cancer dedifferentiation","head and neck squamous cell carcinoma|hnscc|liu2020_immune","head and neck squamous cell carcinoma|hnscc|immune|liu2021_mgs|myeloid cells","head and neck squamous cell carcinoma|hnscc|hpv negative|lohavanichbutr2013_hpvneg|oscc","immune infiltration|lu2020_npc|m6A|mettl3|nasopharyngeal carcinoma|prognostic factors","drug sensitivity|ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis","ifng|interferon gamma|muro2016_ifng","head and neck squamous cell carcinoma|hnscc|immune|qiang2021_irgs","cytolitic activity|immune|rooney2015_cyt ","drug resistance|ferroptosis|pan-cancer|prognosis","immune|she2020_irgs","nasopharyngeal carcinoma|prediction|prognostic factors|progression|si2022_npc","distant metastasis|ebv|epstein-barr virus|nasopharyngeal carcinoma|tang2018_npc","metabolism|wu2020_metabolic","ferroptosis|head and neck squamous cell carcinoma|hnscc|immune microenvironment|prognosis","ca9|ferroptosis|prognosis|thymic squamous cell carcinoma|tscc","immune|nasopharyngeal carcinoma|zou2020_npc"],"publication_doi":["<a href='https://doi.org/10.1172/JCI91190'>10.1172/JCI91190<\/a>","<a href='https://doi.org/10.1155/2019/3865279'>10.1155/2019/3865279<\/a>","<a href='https://doi.org/10.1038/nm.2174'>10.1038/nm.2174<\/a>","<a href='https://doi.org/10.1093/annonc/mdu173'>10.1093/annonc/mdu173<\/a>","<a href='https://doi.org/10.1016/j.ijrobp.2009.06.014'>10.1016/j.ijrobp.2009.06.014<\/a>","<a href='https://doi.org/10.1038/ncomms3612'>10.1038/ncomms3612<\/a>","<a href='https://doi.org/10.1038/ncomms3612'>10.1038/ncomms3612<\/a>","<a href='https://doi.org/10.1158/1078-0432.CCR-13-0542'>10.1158/1078-0432.CCR-13-0542<\/a>","<a href='https://doi.org/10.3389/fgene.2021.732211'>10.3389/fgene.2021.732211<\/a>","<a href='https://doi.org/10.18632/aging.202842'>10.18632/aging.202842<\/a>","<a href='https://doi.org/10.1007/s00405-021-06789-3'>10.1007/s00405-021-06789-3<\/a>","<a href='https://doi.org/10.1016/j.intimp.2021.107789'>10.1016/j.intimp.2021.107789<\/a>","<a href='https://doi.org/10.3389/fcell.2021.739011'>10.3389/fcell.2021.739011<\/a>","<a href='https://doi.org/10.3389/fgene.2020.571403'>10.3389/fgene.2020.571403<\/a>","<a href='https://doi.org/10.1016/j.intimp.2021.108431'>10.1016/j.intimp.2021.108431<\/a>","<a href='https://doi.org/10.1016/j.celrep.2016.12.019'>10.1016/j.celrep.2016.12.019<\/a>","<a href='https://doi.org/10.1016/j.celrep.2016.12.019'>10.1016/j.celrep.2016.12.019<\/a>","<a href='https://doi.org/10.1016/j.celrep.2016.12.019'>10.1016/j.celrep.2016.12.019<\/a>","<a href='https://doi.org/10.1016/j.celrep.2016.12.019'>10.1016/j.celrep.2016.12.019<\/a>","<a href='https://doi.org/10.1186/s12885-021-08478-0'>10.1186/s12885-021-08478-0<\/a>","<a href='https://doi.org/10.21203/rs.3.rs-349666/v2'>10.21203/rs.3.rs-349666/v2<\/a>","<a href='https://doi.org/10.3389/fgene.2021.698040'>10.3389/fgene.2021.698040<\/a>","<a href='https://doi.org/10.2147/IJGM.S343233'>10.2147/IJGM.S343233<\/a>","<a href='https://doi.org/10.1210/clinem/dgab132'>10.1210/clinem/dgab132<\/a>","<a href='https://doi.org/10.3389/fgene.2020.576566'>10.3389/fgene.2020.576566<\/a>","<a href='https://doi.org/10.3389/fimmu.2021.659184'>10.3389/fimmu.2021.659184<\/a>","<a href='https://doi.org/10.1158/1078-0432.CCR-12-2647'>10.1158/1078-0432.CCR-12-2647<\/a>","<a href='https://doi.org/10.3389/fonc.2020.00875'>10.3389/fonc.2020.00875<\/a>","<a href='https://doi.org/10.3389/fgene.2021.755486'>10.3389/fgene.2021.755486<\/a>","<a href='https://doi.org/10.1016/S1470-2045(16)00175-3'>10.1016/S1470-2045(16)00175-3<\/a>","<a href='https://doi.org/10.1016/j.csbj.2021.01.046'>10.1016/j.csbj.2021.01.046<\/a>","<a href='https://doi.org/10.1016/j.cell.2014.12.033'>10.1016/j.cell.2014.12.033<\/a>","<a href='https://doi.org/10.3389/fcell.2021.794475'>10.3389/fcell.2021.794475<\/a>","<a href='https://doi.org/10.1186/s12935-020-1104-7'>10.1186/s12935-020-1104-7<\/a>","<a href='https://doi.org/10.1177/15330338221080972'>10.1177/15330338221080972<\/a>","<a href='https://doi.org/10.1016/S1470-2045(18)30080-9'>10.1016/S1470-2045(18)30080-9<\/a>","<a href='https://doi.org/10.1155/2020/6716908'>10.1155/2020/6716908<\/a>","<a href='https://doi.org/10.1155/2021/5759927'>10.1155/2021/5759927<\/a>","<a href='https://doi.org/10.1002/2211-5463.13348'>10.1002/2211-5463.13348<\/a>","<a href='https://doi.org/10.1016/j.lfs.2020.118620'>10.1016/j.lfs.2020.118620<\/a>"],"description":["Immune expanded gene signature aimed at predicting response to pembrolizumab in multiple cancer types.","Immune/inflammatory-related genes for prognostic prediction of oral cavity squamous cell carcinoma.","Biomarker aimed at identifying high-risk soft-tissue sarocmas and predicting metastasis outcome.","Signature aimed at stratifying HNSCC patients in high/low risk of relapse.","Genes aimed at predicting radiosensitivity. A higher index correspond to a more radioresistant patient.","Immune-related genes used in the Estimation of STromal and Immune cells in MAlignant Tumours using Expression data (ESTIMATE) method.","Stroma-related genes used in the Estimation of STromal and Immune cells in MAlignant Tumours using Expression data (ESTIMATE) method.","Hypoxia-related genes aimed at predicting benefit from hypoxia-modifying treatment in laryngeal cancer.","A model of ferroptosis-related genes which can be performed in clinical applications, used to guide immunotherapy and chemotherapy and which can predict the prognosis of HNSCC patients","Immune-related genes for survival prediction in HNSCC patients.","A ferroptosis-related model aimed at predicting the prognosis of laryngeal cancer and for stratifying patients into low- and high-risk groups","authors generate and validate in TCGA a prognostic signature to predict prognosis and immune microenvironment in HNSCC","This study aimed to investigate distinct ferroptosis-related genes, their prognostic value and their relationship with immunotherapy in patients with head and neck squamous cell carcinoma (HNSCC).","Prognostic signature of genes related to RNA-binding proteins for HNSCC.","The authors designed and validated a seven ferroptosis-associated gene signature (FGS) and further probed the potential survival mechanisms along with therapeutic targets for HNSCC","One of the four categories composing the Immunophenoscore: co-inhibitory and co-stimulatory biomarkers.","One of the four categories composing the Immunophenoscore: infiltration of activated CD8+/CD4+ T cells and Tem CD8+/CD4+ cells.","One of the four categories composing the Immunophenoscore: MHC class I/II and non-classical biomarkers.","One of the four categories composing the Immunophenoscore: infiltration of immunosuppressive cells (Tregs and MDSCs).","A 10-ferroptosis-releated gene signature that can be used to predict the prognosis of OSCC patients for future therapies based on potential FRG targets.","A prognostic model of HNSCC based on ferroptosis-related genes","A prognostic model based on FRGs related to OS providing a new molecular target for the prediction of HNSCC prognosis","A six-ferroptosis-related-gene model crossing clinical prognostic parameters that can predict HNSCC patient prognosis and provide a reliable prognostic evaluation","Immune-related genes associated to dedifferentiation and immune exhaustion in thyroid cancer.","Immune-related genes associated to overall survival in HNSCC patients.","Prognostic signature of myeloid-related genes for HNSCC.","Signature prognostic of HPV-negative oral squamous cell carcinoma.","A prognostic signature related to m6A genes in nasopharyngeal carcinoma.","A ferroptosis-related prognostic signature correlated with sensitivity of common chemotherapy drugs for HNSCC and associated with drug sensitivity to cancer cells","Interferon gamma-related genes found to be significantly associated with clinical response to pembrolizumab in patients with melanoma.","Prognostic signature of immune-related genes for HNSCC.","Genes giving a measure of immune cytolytic activity in tumors.","A 15-signature model based on the gene expression data of cancer cell lines; the validation cohorts showed higher expression levels of 12 ferroptosis-related genes in HS groups","Prognostic signature of immune-related genes for HNSCC.","A prognostic signature to predict patients with distant metastasis in nasopharyngeal carcinoma","A prognostic signature to predict patients with distant metastasis in locoregionally advanced nasopharyngeal carcinoma.","Metabolic prognostic signature for HNSCC patients.","An FRG prognostic signature model for HNSCC","Potential diagnostic biomarkers associated with TSCC","A prognostic signature of tumor-infiltrating immune cells in nasopharyngeal carcinoma"]},"columns":[{"id":"signature_id","name":"signature ID","type":"character"},{"id":"signature_keywords","name":"signature_keywords","type":"character","filterable":true,"searchable":true,"show":false},{"id":"publication_doi","name":"DOI","type":"character","html":true},{"id":"description","name":"description","type":"character"}],"searchable":true,"defaultPageSize":10,"highlight":true,"striped":true,"dataKey":"6780d357d3367bae7d7af38c25af0267"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script></main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrea Carenzo, Loris De Cecco, Federico Pistore.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
