<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="hacksig">
<title>Introduction to hacksig • hacksig</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to hacksig">
<meta property="og:description" content="hacksig">
<meta property="og:image" content="https://acare.github.io/hacksig/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">hacksig</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/hacksig.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/available-sigs.html">Available signatures</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Acare/hacksig/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introduction to hacksig</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Acare/hacksig/blob/HEAD/vignettes/hacksig.Rmd" class="external-link"><code>vignettes/hacksig.Rmd</code></a></small>
      <div class="d-none name"><code>hacksig.Rmd</code></div>
    </div>

    
    
<p>Hacksig is a collection of cancer transcriptomics gene signatures and
it provides a simple and tidy interface to compute single sample
enrichment scores.</p>
<p>This document will show you how to getting started with hacksig, but
first, we must load the following packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Acare/hacksig" class="external-link">hacksig</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to plot and transform data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to get the MSigDB gene signatures</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igordot.github.io/msigdbr/" class="external-link">msigdbr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to parallelize computations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="available-signatures-from-the-literature">Available signatures from the literature<a class="anchor" aria-label="anchor" href="#available-signatures-from-the-literature"></a>
</h2>
<p>In order to get a complete list of the implemented signatures, you
can use <code><a href="../reference/get_sig_info.html">get_sig_info()</a></code>. It returns a tibble with very
useful information:</p>
<ul>
<li>the <code>signature_id</code>;</li>
<li>a string of keywords associated to a signature (separated by the
“pipe” <code>|</code> symbol);</li>
<li>the <code>publication_doi</code> linking to the original
publication;</li>
<li>a brief <code>description</code>.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_sig_info.html">get_sig_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 40 × 4</span></span></span>
<span><span class="co">#&gt;   signature_id       signature_keywords                          publi…¹ descr…²</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ayers2017_immexp   ayers2017_immexp|immune expanded            10.117… Immune…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> bai2019_immune     bai2019_immune|head and neck squamous cell… 10.115… Immune…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> cinsarc            cinsarc|metastasis|sarcoma|sts              10.103… Biomar…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> dececco2014_int172 dececco2014_int172|head and neck squamous … 10.109… Signat…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> eschrich2009_rsi   eschrich2009_rsi|radioresistance|radiosens… 10.101… Genes …</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 35 more rows, and abbreviated variable names ¹​publication_doi,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ²​description</span></span></span></code></pre></div>
<p>If you want to get the list of gene symbols for one or more of the
implemented signatures, then use <code><a href="../reference/get_sig_genes.html">get_sig_genes()</a></code> with valid
keywords:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_sig_genes.html">get_sig_genes</a></span><span class="op">(</span><span class="st">"ifng"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $muro2016_ifng</span></span>
<span><span class="co">#&gt; [1] "CXCL10"  "CXCL9"   "HLA-DRA" "IDO1"    "IFNG"    "STAT1"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="check-your-signatures">Check your signatures<a class="anchor" aria-label="anchor" href="#check-your-signatures"></a>
</h2>
<p>The first thing you should do before computing scores for a signature
is to check how many of its genes are present in your data. To
accomplish this, we can use <code><a href="../reference/check_sig.html">check_sig()</a></code> on a normalized
gene expression matrix (either microarray or RNA-seq normalized data),
which must be formatted as an object of class <code>matrix</code> or
<code>data.frame</code> with gene symbols as row names and sample IDs as
column names.</p>
<p>For this tutorial, we will use <code>test_expr</code> (an R object
included in hacksig) as an example gene expression matrix with 20
simulated samples.</p>
<p>By default, <code><a href="../reference/check_sig.html">check_sig()</a></code> will compute statistics for
every signature implemented in <code>hacksig</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_sig.html">check_sig</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 40 × 5</span></span></span>
<span><span class="co">#&gt;   signature_id     n_genes n_present frac_present missing_genes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> wu2020_metabolic      30        20        0.667 <span style="color: #949494;">&lt;chr [10]&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> muro2016_ifng          6         4        0.667 <span style="color: #949494;">&lt;chr [2]&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> liu2020_immune         6         4        0.667 <span style="color: #949494;">&lt;chr [2]&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> liu2021_mgs            6         4        0.667 <span style="color: #949494;">&lt;chr [2]&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> lu2020_npc             3         2        0.667 <span style="color: #949494;">&lt;chr [1]&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 35 more rows</span></span></span></code></pre></div>
<p>You can filter for specific signatures by entering keywords in the
<code>signatures</code> argument (partial matching and regular
expressions will work too):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_sig.html">check_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, signatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"metab"</span>, <span class="st">"cinsarc"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 5</span></span></span>
<span><span class="co">#&gt;   signature_id     n_genes n_present frac_present missing_genes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> wu2020_metabolic      30        20        0.667 <span style="color: #949494;">&lt;chr [10]&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> cinsarc               67        40        0.597 <span style="color: #949494;">&lt;chr [27]&gt;</span></span></span></code></pre></div>
<p>We can also check for signatures not implemented in hacksig, that is
custom signatures. For example, we can use the <code>msigdbr</code>
package to download the <em>Hallmark</em> gene set collection as a
tibble and transform it into a list:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmark_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">gene_symbol</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">gene_symbol</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">genes</span>, <span class="fu"><a href="https://purrr.tidyverse.org/reference/compose.html" class="external-link">compose</a></span><span class="op">(</span><span class="va">as_vector</span>, <span class="va">unname</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">deframe</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/check_sig.html">check_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">hallmark_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 50 × 5</span></span></span>
<span><span class="co">#&gt;   signature_id                        n_genes n_present frac_present missing_g…¹</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> HALLMARK_WNT_BETA_CATENIN_SIGNALING      42        27        0.643 <span style="color: #949494;">&lt;chr [15]&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> HALLMARK_APICAL_SURFACE                  44        28        0.636 <span style="color: #949494;">&lt;chr [16]&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> HALLMARK_BILE_ACID_METABOLISM           112        70        0.625 <span style="color: #949494;">&lt;chr [42]&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> HALLMARK_NOTCH_SIGNALING                 32        20        0.625 <span style="color: #949494;">&lt;chr [12]&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> HALLMARK_PI3K_AKT_MTOR_SIGNALING        105        65        0.619 <span style="color: #949494;">&lt;chr [40]&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 45 more rows, and abbreviated variable name ¹​missing_genes</span></span></span></code></pre></div>
<p>Missing genes for the <code>HALLMARK_NOTCH_SIGNALING</code> gene set
are:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_sig.html">check_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">hallmark_list</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">signature_id</span> <span class="op">==</span> <span class="st">"HALLMARK_NOTCH_SIGNALING"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">missing_genes</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;  [1] "FZD5"    "HEYL"    "KAT2A"   "MAML2"   "NOTCH1"  "NOTCH3"  "PPARD"  </span></span>
<span><span class="co">#&gt;  [8] "PRKCA"   "PSEN2"   "SAP30"   "ST3GAL6" "TCF7L2"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compute-single-sample-scores">Compute single sample scores<a class="anchor" aria-label="anchor" href="#compute-single-sample-scores"></a>
</h2>
<div class="section level3">
<h3 id="hack_sig">hack_sig<a class="anchor" aria-label="anchor" href="#hack_sig"></a>
</h3>
<p>The main function of the package, <code><a href="../reference/hack_sig.html">hack_sig()</a></code>, permits to
obtain single sample scores from gene signatures. By default, it will
compute scores for all the signatures implemented in the package with
the original publication method.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: <span style="color: #0000BB;">ℹ</span> No genes are present in 'expr_data' for the following signatures:</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> zhu2021_ferroptosis</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> rooney2015_cyt</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> To obtain CINSARC, ESTIMATE and Immunophenoscore with the original procedures, see:</span></span>
<span><span class="co">#&gt; ?hack_cinsarc</span></span>
<span><span class="co">#&gt; ?hack_estimate</span></span>
<span><span class="co">#&gt; ?hack_immunophenoscore</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 32</span></span></span>
<span><span class="co">#&gt;   sample_id ayers2017_…¹ bai20…² decec…³ eschr…⁴ eusta…⁵ fan20…⁶ fang2…⁷ han20…⁸</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1           5.71   -<span style="color: #BB0000;">22.3</span>    2.62  0.028<span style="text-decoration: underline;">9</span>    8.1   -<span style="color: #BB0000;">2.67</span>     2.96  -<span style="color: #BB0000;">0.503</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample10          6.96   -<span style="color: #BB0000;">23.0</span>    2.69  0.415     8.17  -<span style="color: #BB0000;">0.743</span>    3.42  -<span style="color: #BB0000;">0.259</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample11          8.06   -<span style="color: #BB0000;">19.9</span>    1.73  0.542     7.79  -<span style="color: #BB0000;">5.25</span>     4.01  -<span style="color: #BB0000;">0.454</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample12          8.57   -<span style="color: #BB0000;">23.8</span>    2.35  0.287     8.45  -<span style="color: #BB0000;">2.96</span>     8.45  -<span style="color: #BB0000;">0.450</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample13          6.35   -<span style="color: #BB0000;">25.9</span>    2.36  0.583     8.22  -<span style="color: #BB0000;">4.52</span>     1.97  -<span style="color: #BB0000;">0.317</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows, 23 more variables: he2021_ferroptosis_a &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   he2021_ferroptosis_b &lt;dbl&gt;, hu2021_derbp &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   huang2022_ferroptosis &lt;dbl&gt;, li2021_ferroptosis_a &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   li2021_ferroptosis_b &lt;dbl&gt;, li2021_ferroptosis_c &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   li2021_ferroptosis_d &lt;dbl&gt;, li2021_irgs &lt;dbl&gt;, liu2020_immune &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   liu2021_mgs &lt;dbl&gt;, lohavanichbutr2013_hpvneg &lt;dbl&gt;, lu2020_npc &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   lu2021_ferroptosis &lt;dbl&gt;, muro2016_ifng &lt;dbl&gt;, qiang2021_irgs &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>You can also filter for specific signatures (e.g. the immune and
stromal ESTIMATE signatures) and choose a particular single sample
method:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, signatures <span class="op">=</span> <span class="st">"estimate"</span>, method <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 3</span></span></span>
<span><span class="co">#&gt;   sample_id estimate_immune estimate_stromal</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1            -<span style="color: #BB0000;">2.65</span>            -<span style="color: #BB0000;">0.262</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample10            1.37             0.305</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample11            1.50            -<span style="color: #BB0000;">0.959</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample12            1.65            -<span style="color: #BB0000;">1.22</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample13           -<span style="color: #BB0000;">0.535</span>           -<span style="color: #BB0000;">0.743</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows</span></span></span></code></pre></div>
<p>Valid choices for single sample <code>method</code>s are:</p>
<ul>
<li>
<code>"zscore"</code>, for the combined z-score;</li>
<li>
<code>"ssgsea"</code>, for the single sample GSEA;</li>
<li>
<code>"singscore"</code>, for the singscore method.</li>
</ul>
<p>Run <code><a href="../reference/hack_sig.html">?hack_sig</a></code> to see additional parameter specifications
for these methods.</p>
<p>As in <code><a href="../reference/check_sig.html">check_sig()</a></code>, the argument <code>signatures</code>
can also be a list of gene signatures. For example, we can compute
normalized single sample GSEA scores for the Hallmark gene sets:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">hallmark_list</span>, </span>
<span>         method <span class="op">=</span> <span class="st">"ssgsea"</span>, sample_norm <span class="op">=</span> <span class="st">"separate"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 51</span></span></span>
<span><span class="co">#&gt;   sample_id HALLMARK_A…¹ HALLM…² HALLM…³ HALLM…⁴ HALLM…⁵ HALLM…⁶ HALLM…⁷ HALLM…⁸</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1          0.683   0.419   0.943   0.447   0.137  0.052<span style="text-decoration: underline;">9</span>   0.484   0.454</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample10         0.384   0.790   0.300   0.287   0.373  0.290    0.175   0.482</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample11         0.249   0.756   0.646   1       0.490  0        0.377   0.676</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample12         0.998   1       0.959   0.111   1      0.380    0.452   0.515</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample13         0.785   0       0.373   0.607   0.398  0.309    0.217   0.274</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows, 42 more variables:</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_CHOLESTEROL_HOMEOSTASIS &lt;dbl&gt;, HALLMARK_COAGULATION &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_COMPLEMENT &lt;dbl&gt;, HALLMARK_DNA_REPAIR &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_E2F_TARGETS &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_ESTROGEN_RESPONSE_EARLY &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_ESTROGEN_RESPONSE_LATE &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>There are three methods for which <code><a href="../reference/hack_sig.html">hack_sig()</a></code> cannot be
used to compute gene signature scores with the original method. These
are: CINSARC, ESTIMATE and the Immunophenoscore.</p>
</div>
<div class="section level3">
<h3 id="hack_cinsarc">hack_cinsarc<a class="anchor" aria-label="anchor" href="#hack_cinsarc"></a>
</h3>
<p>For the CINSARC classification, you must provide a vector with
distant metastasis status:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">rand_dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hack_cinsarc.html">hack_cinsarc</a></span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">rand_dm</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 2</span></span></span>
<span><span class="co">#&gt;   sample_id cinsarc_class</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1   C2           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample2   C1           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample3   C2           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample4   C1           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample5   C2           </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="hack_estimate">hack_estimate<a class="anchor" aria-label="anchor" href="#hack_estimate"></a>
</h3>
<p>Immune, stromal, ESTIMATE and tumor purity scores from the ESTIMATE
method can be obtained with:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hack_estimate.html">hack_estimate</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 5</span></span></span>
<span><span class="co">#&gt;   sample_id immune_score stroma_score estimate_score purity_score</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1          -<span style="color: #BB0000;">636.</span>         778.           142.        0.811</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample10         <span style="text-decoration: underline;">1</span>590.        <span style="text-decoration: underline;">1</span>297.          <span style="text-decoration: underline;">2</span>887.        0.516</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample11         <span style="text-decoration: underline;">2</span>040.         512.          <span style="text-decoration: underline;">2</span>552.        0.557</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample12         <span style="text-decoration: underline;">1</span>835.         772.          <span style="text-decoration: underline;">2</span>607.        0.551</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample13          632.         778.          <span style="text-decoration: underline;">1</span>409.        0.688</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="hack_immunophenoscore">hack_immunophenoscore<a class="anchor" aria-label="anchor" href="#hack_immunophenoscore"></a>
</h3>
<p>Finally, the raw immunophenoscore and its discrete (0-10 normalized)
counterpart can be obtained with:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hack_immunophenoscore.html">hack_immunophenoscore</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 3</span></span></span>
<span><span class="co">#&gt;   sample_id raw_score ips_score</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1      0.942          3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample2     -<span style="color: #BB0000;">0.348</span>          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample3      0.093<span style="text-decoration: underline;">9</span>         0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample4     -<span style="color: #BB0000;">0.335</span>          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample5      1.64           5</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows</span></span></span></code></pre></div>
<p>You can also obtain all biomarker scores with:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hack_immunophenoscore.html">hack_immunophenoscore</a></span><span class="op">(</span><span class="va">test_expr</span>, extract <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 19</span></span></span>
<span><span class="co">#&gt;   sample_id act_cd4_sc…¹ act_c…² b2m_s…³ cd27_…⁴ icos_…⁵ mdsc_…⁶ pd1_s…⁷ pdl2_…⁸</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1         0.079<span style="text-decoration: underline;">3</span> -<span style="color: #BB0000;">0.335</span>    0.312  -<span style="color: #BB0000;">0.868</span>  -<span style="color: #BB0000;">0.768</span>  0.652    1.66   -<span style="color: #BB0000;">0.674</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample2        -<span style="color: #BB0000;">0.016</span><span style="color: #BB0000; text-decoration: underline;">5</span> -<span style="color: #BB0000;">0.308</span>   -<span style="color: #BB0000;">0.796</span>   0.195   0.909 -<span style="color: #BB0000;">0.273</span>   -<span style="color: #BB0000;">1.71</span>    0.304</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample3         0.539   0.039<span style="text-decoration: underline;">3</span>  -<span style="color: #BB0000;">0.483</span>   0.525  -<span style="color: #BB0000;">0.457</span>  0.150    1.38    1.18 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample4         0.398   0.383   -<span style="color: #BB0000;">0.777</span>   0.699   0.658  0.130    0.101  -<span style="color: #BB0000;">0.168</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample5        -<span style="color: #BB0000;">0.049</span><span style="color: #BB0000; text-decoration: underline;">6</span>  0.137    0.876   0.373   0.251  0.041<span style="text-decoration: underline;">3</span>   2.70   -<span style="color: #BB0000;">0.113</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows, 10 more variables: tem_cd4_score &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   tem_cd8_score &lt;dbl&gt;, tigit_score &lt;dbl&gt;, treg_score &lt;dbl&gt;, raw_score &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ips_score &lt;dbl&gt;, cp_score &lt;dbl&gt;, ec_score &lt;dbl&gt;, mhc_score &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sc_score &lt;dbl&gt;, and abbreviated variable names ¹​act_cd4_score,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ²​act_cd8_score, ³​b2m_score, ⁴​cd27_score, ⁵​icos_score, ⁶​mdsc_score,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ⁷​pd1_score, ⁸​pdl2_score</span></span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="stratify-your-samples">Stratify your samples<a class="anchor" aria-label="anchor" href="#stratify-your-samples"></a>
</h2>
<p>If you want to categorize your samples into two or more signature
classes based on a score cutoff, you can use <code><a href="../reference/stratify_sig.html">stratify_sig()</a></code>
after <code><a href="../reference/hack_sig.html">hack_sig()</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_expr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="st">"estimate"</span>, method <span class="op">=</span> <span class="st">"singscore"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/stratify_sig.html">stratify_sig</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 3</span></span></span>
<span><span class="co">#&gt;   sample_id estimate_immune estimate_stromal</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1   low             low             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample10  high            high            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample11  high            low             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample12  high            low             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample13  low             low             </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows</span></span></span></code></pre></div>
<p>By default, <code><a href="../reference/stratify_sig.html">stratify_sig()</a></code> will stratify samples either
with the original publication method (if any) or by the median score
(otherwise). <code><a href="../reference/stratify_sig.html">stratify_sig()</a></code> will work only with signatures
implemented in <code>hacksig</code>.</p>
</div>
<div class="section level2">
<h2 id="speed-up-computation-time">Speed-up computation time<a class="anchor" aria-label="anchor" href="#speed-up-computation-time"></a>
</h2>
<p>Our rank-based single sample method implementations (i.e. single
sample GSEA and singscore) are slower than their counterparts
implemented in <code>GSVA</code> and <code>singscore</code>. Hence, to
speed-up computation time you can use the <code>future</code>
package:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">hallmark_list</span>, method <span class="op">=</span> <span class="st">"ssgsea"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 51</span></span></span>
<span><span class="co">#&gt;   sample_id HALLMARK_A…¹ HALLM…² HALLM…³ HALLM…⁴ HALLM…⁵ HALLM…⁶ HALLM…⁷ HALLM…⁸</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample1          <span style="text-decoration: underline;">1</span>593.    709.  <span style="text-decoration: underline;">2</span>013.   <span style="text-decoration: underline;">1</span>396.     378. -<span style="color: #BB0000; text-decoration: underline;">1</span><span style="color: #BB0000;">299.</span>    <span style="text-decoration: underline;">1</span>578.    910.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample10          739.   <span style="text-decoration: underline;">1</span>476.    37.4   882.     793.   -<span style="color: #BB0000;">25.6</span>    873.   <span style="text-decoration: underline;">1</span>015.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample11          572.   <span style="text-decoration: underline;">1</span>497.  <span style="text-decoration: underline;">1</span>083.   <span style="text-decoration: underline;">2</span>731.     917. -<span style="color: #BB0000; text-decoration: underline;">1</span><span style="color: #BB0000;">430.</span>    <span style="text-decoration: underline;">1</span>294.   <span style="text-decoration: underline;">1</span>588.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample12         <span style="text-decoration: underline;">2</span>426.   <span style="text-decoration: underline;">1</span>964.  <span style="text-decoration: underline;">2</span>168.     19.0   <span style="text-decoration: underline;">1</span>849.   693.    <span style="text-decoration: underline;">1</span>668.   <span style="text-decoration: underline;">1</span>137.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample13         <span style="text-decoration: underline;">1</span>822.    101.   166.   <span style="text-decoration: underline;">1</span>659.    <span style="text-decoration: underline;">1</span>141.   153.     969.    378.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 15 more rows, 42 more variables:</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_CHOLESTEROL_HOMEOSTASIS &lt;dbl&gt;, HALLMARK_COAGULATION &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_COMPLEMENT &lt;dbl&gt;, HALLMARK_DNA_REPAIR &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_E2F_TARGETS &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_ESTROGEN_RESPONSE_EARLY &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   HALLMARK_ESTROGEN_RESPONSE_LATE &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="use-case">Use case<a class="anchor" aria-label="anchor" href="#use-case"></a>
</h2>
<p>Let’s say we want to compute single sample scores for the KEGG gene
set collection and then correlate these scores with the tumor purity
given by the ESTIMATE method.</p>
<p>First, we get the KEGG list and use <code><a href="../reference/check_sig.html">check_sig()</a></code> to keep
only those gene sets whose genes are more than 2/3 present in our gene
expression matrix.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, subcategory <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">gene_symbol</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">gene_symbol</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">genes</span>, <span class="fu"><a href="https://purrr.tidyverse.org/reference/compose.html" class="external-link">compose</a></span><span class="op">(</span><span class="va">as_vector</span>, <span class="va">unname</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">deframe</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kegg_ok</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_sig.html">check_sig</a></span><span class="op">(</span><span class="va">test_expr</span>, <span class="va">kegg_list</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">frac_present</span> <span class="op">&gt;</span> <span class="fl">0.66</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">signature_id</span><span class="op">)</span></span></code></pre></div>
<p>Then, we apply both the combined z-score and the ssGSEA method for
the resulting list of 10 KEGG gene sets using
<code><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">purrr::map_dfr()</a></code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>zscore <span class="op">=</span> <span class="st">"zscore"</span>, ssgsea <span class="op">=</span> <span class="st">"ssgsea"</span><span class="op">)</span>, </span>
<span>                       <span class="op">~</span> <span class="fu"><a href="../reference/hack_sig.html">hack_sig</a></span><span class="op">(</span><span class="va">test_expr</span>,</span>
<span>                                  <span class="va">kegg_list</span><span class="op">[</span><span class="va">kegg_ok</span><span class="op">]</span>,</span>
<span>                                  method <span class="op">=</span> <span class="va">.x</span>,</span>
<span>                                  sample_norm <span class="op">=</span> <span class="st">"separate"</span><span class="op">)</span>,</span>
<span>                       .id <span class="op">=</span> <span class="st">"method"</span><span class="op">)</span></span></code></pre></div>
<p>We can transform the <code>kegg_scores</code> tibble in long format
using <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">tidyr::pivot_longer()</a></code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_scores</span> <span class="op">&lt;-</span> <span class="va">kegg_scores</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"KEGG"</span><span class="op">)</span>, </span>
<span>                 names_to <span class="op">=</span> <span class="st">"kegg_id"</span>, values_to <span class="op">=</span> <span class="st">"kegg_score"</span><span class="op">)</span></span></code></pre></div>
<p>Finally, after computing the tumor purity scores, we can merge the
two data sets and plot the results:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">purity_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hack_estimate.html">hack_estimate</a></span><span class="op">(</span><span class="va">test_expr</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">purity_score</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kegg_scores</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">purity_scores</span>, by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">kegg_id</span>, y <span class="op">=</span> <span class="va">kegg_score</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">purity_score</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">method</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"enrichment score"</span>, color <span class="op">=</span> <span class="st">"Tumor purity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span></code></pre></div>
<p><img src="hacksig_files/figure-html/unnamed-chunk-21-1.png" width="864"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Andrea Carenzo, Loris De Cecco, Federico Pistore.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
